Swift Concurrency adoption guidelines for Swift Server Libraries (#70)
swift-server/guides: docs/concurrency-adoption-guidelines.md

> # Swift Concurrency adoption guidelines for Swift Server Libraries

Reviewer: may be nice to add an "index" at the end before we merge so people can see the high level outline/jump to particular sections as this is getting rather lengthy

Reviewer 2: Good point, will do

> This writeup attempts to provide a set of guidelines to follow by authors of server-side swift libraries. Specifically a lot of the discussion here revolves around what to do about existing APIs and libraries making extensive use of Swift NIO’s EventLoopFuture and related types.

Reviewer: ```suggestion
This writeup attempts to provide a set of guidelines to follow by authors of server-side Swift libraries. Specifically a lot of the discussion here revolves around what to do about existing APIs and libraries making extensive use of Swift NIO’s `EventLoopFuture` and related types.
```

Swift Concurrency is a multi-year effort. It is very valuable for the server community to participate in this multi-year adoption of the concurrency features, one by one, and provide feedback while doing so. As such, we should not hold off adopting concurrency features until Swift 6 as we may miss valuable opportunity to improve the concurrency model.

> In 2021 we saw structured concurrency and actors. Now is a great time to provide APIs using those primitives. In the future we will see fully checked Swift concurrency. This will come with breaking changes. For this reason adopting the new concurrency features can be split into two phases.

Reviewer: ```suggestion
In 2021 we saw structured concurrency and actors arrive with Swift 5.5. Now is a great time to provide APIs using those primitives. In the future we will see fully checked Swift concurrency. This will come with breaking changes. For this reason adopting the new concurrency features can be split into two phases.
```

## What you can do right now

### API Design

> Firstly, existing libraries should strive to add `async` functions where possible to their user-facing “surface” APIs in addition to existing `*Future` based APIs wherever possible. These additive APIs can be gated on the Swift version and can be added without breaking existing users, for example like this:

Reviewer: do we expect anyone who this guide applies to is using any future type besides `EventLoopFuture`?

Reviewer: ```suggestion
Firstly, existing libraries should strive to add `async` functions where possible to their user-facing “surface” APIs in addition to existing `*Future` based APIs wherever possible. These additive APIs can be gated on the Swift version and can be added without breaking existing users' code, for example like this:
```

Reviewer 2: Hmm... I guess there MIGHT BE some library using the other future libraries but honestly I've not seen one for server focus that would use the other ones (PromiseKit etc), they're usually a bit iOS "tainted" hm...

```swift
extension Worker {
  func work() -> EventLoopFuture<Value> { ... }
  
  #if swift(>=5.5)
  func work() async throws -> Value { ... }
  #endif
}
```

> If a function cannot fail but was using futures before, it should not include the throws keyword in its new incarnation. 

Reviewer: ```suggestion
If a function cannot fail but was using futures before, it should not include the `throws` keyword in its new incarnation. 
```

> If a function cannot fail but was using futures before, it should not include the `throws` keyword in its new incarnation. 

Reviewer 3: It's possible existing APIs guarantee the future will never fail by always `recover`-ing from an error (i.e. `try!` is fine) so I think this point is still valid.

Such adoption can begin immediately, and should not cause any issues to existing users of existing libraries. 

### SwiftNIO helper functions

> To allow an easy transition to async code, SwiftNIO offers a number of helper methods on EventLoopFuture and -Promise. Those live in the `_NIOConcurrency` module and will move to NIOCore once Swift concurrency is released.

Reviewer: ```suggestion
To allow an easy transition to async code, SwiftNIO offers a number of helper methods on `EventLoopFuture` and `-Promise`. Those live in the `_NIOConcurrency` module and will move to `NIOCore` once Swift concurrency is released.
```

> To allow an easy transition to async code, SwiftNIO offers a number of helper methods on `EventLoopFuture` and `-Promise`. Those live in the `_NIOConcurrency` module and will move to `NIOCore` once Swift concurrency is released.

Reviewer 2: No, it'll remove it. The guide will reflect that once the team changes the module name :)

> On every `EventLoopFuture` you can call `.get()` to transition the future into an awaitable invocation. If you want to translate async/await calls to an EventLoopFuture we recommend the following pattern:

Reviewer: ```suggestion
On every `EventLoopFuture` you can call `.get()` to transition the future into an `await`-able invocation. If you want to translate async/await calls to an `EventLoopFuture` we recommend the following pattern:
```

```swift 
#if swift(>=5.5)
func yourAsyncFunctionConvertedToAFuture(on eventLoop: EventLoop) 
    -> EventLoopFuture<Result> {

>     let promise = context.eventLoop.makePromise(of: Out.self)

Reviewer: the return type of the method is `EventLoopFuture<Result>` but here we make an `EventLoopPromise<Out>`, should the generic types match?

Reviewer 2: Oh, thanks 👍

Reviewer 4: I think you don't need `context.` because `eventLoop` is passed directly as a function parameter.

    promise.completeWithTask {
        try await yourMethod(yourInputs)
    }
    return promise.futureResult
}
#endif
```

Further helpers exist for `EventLoopGroup`, `Channel`, `ChannelOutboundInvoker` and `ChannelPipeline`.

### Sendable Checking

> > [SE-0302][SE-0302] introduced the Sendable protocol, which is used to indicate which types have values that can safely be copied across actors or, more generally, into any context where a copy of the value might be used concurrently with the original. Applied uniformly to all Swift code, Sendable checking eliminates a large class of data races caused by shared mutable state.

Reviewer: ```suggestion
> [SE-0302][SE-0302] introduced the `Sendable` protocol, which is used to indicate which types have values that can safely be copied across actors or, more generally, into any context where a copy of the value might be used concurrently with the original. Applied uniformly to all Swift code, `Sendable` checking eliminates a large class of data races caused by shared mutable state.
```

>

> > -- from [Staging in Sendable checking][sendable-staging], which outlines the Sendable adoption plan for Swift 6.

Reviewer: ```suggestion
> -- from [Staging in Sendable checking][sendable-staging], which outlines the `Sendable` adoption plan for Swift 6.
```

In the future we will see fully checked Swift concurrency. The language features to support this are the `Sendable` protocol and the `@Sendable` keyword for closures. Since sendable checking will break existing Swift code, a new major Swift version is required.

> To easy the transition to fully checked Swift code, it is possible to annotate your APIs with the Sendable protocol today.

Reviewer 5: "To easy" -> "To ease"

> To ease the transition to fully checked Swift code, it is possible to annotate your APIs with the Sendable protocol today.

Reviewer: ```suggestion
To ease the transition to fully checked Swift code, it is possible to annotate your APIs with the `Sendable` protocol today.
```

You can start adopting Sendable and getting appropriate warnings in Swift 5.5 already by passing the `-warn-concurrency` flag, you can do so in SwiftPM for the entire project like so:

```
swift build -Xswiftc -Xfrontend -Xswiftc -warn-concurrency
```

#### Sendable checking today

Sendable checking is currently disabled in Swift 5.5(.0) because it was causing a number of tricky situations for which we lacked the tools to resolve.

> Most of these issues have been resolved on today’s `main` branch of the compiler, and are expected to land in the next Swift 5.5 releases. It may be worthwhile waiting with adoption until the next version(s) after 5.5.0.

Reviewer: ```suggestion
Most of these issues have been resolved on today’s `main` branch of the compiler, and are expected to land in the next Swift 5.5 releases. It may be worthwhile waiting for adoption until the next version(s) after 5.5.0.
```

> For example, one of such capabilities is the ability for tuples of `Sendable` types to conform to `Sendable` as well. We recommend holding off adoption of `Sendable` until this patch lands in Swift 5.5 (which should be relatively soon). With this change, the difference between Swift 5.5 with `-warn-concurrency` enabled and Swift 6 mode should be very small, and manageable on a case by case basis.

Reviewer 6: It would be better to put an expected date in rather than relatively soon or this guide will not read well in a few months time.

#### Backwards compatibility of declarations and “checked” Swift Concurrency

Adopting Swift Concurrency will progressively cause more warnings, and eventually compile time errors in Swift 6 when sendability checks are violated, marking potentially unsafe code.

> It may be difficult for a library to maintain a “pre Swift 6” version along with a Swift 6 version that fully embraces the new concurrency checks. For example, it may be necessary to mark generic types as Sendable, like so:

Reviewer: ```suggestion
It may be difficult for a library to maintain a version that is compatible with versions prior to Swift 6 while also fully embracing the new concurrency checks. For example, it may be necessary to mark generic types as `Sendable`, like so:
```

```swift
struct Container<Value: Sendable>: Sendable { ... }
```

> Here, the `Value` type must be marked `Sendable` for Swift 6’s concurrency checks to work properly with such container. However, since the Sendable type does not exist in Swift prior to Swift 5.5, it would be difficult to maintain a library that supports both Swift 5.4+ as well as Swift 6.

Reviewer: ```suggestion
Here, the `Value` type must be marked `Sendable` for Swift 6’s concurrency checks to work properly with such container. However, since the `Sendable` type does not exist in Swift prior to Swift 5.5, it would be difficult to maintain a library that supports both Swift 5.4+ as well as Swift 6.
```

> Here, the Value type must be marked Sendable for Swift 6’s concurrency checks to be able property work with such container. However, since the Sendable type does not exist in Swift prior to Swift 5.5, it would be difficult to maintain a library that supports both Swift 5.4+ as well as Swift 6.

Reviewer 5: "to be able property work" -> "to be able to properly work" or simply "to properly work"

> In such situations, it may be helpful to utilize the following trick to be able to share the same Container declaration between both Swift versions of the library:

Reviewer: ```suggestion
In such situations, it may be helpful to utilize the following trick to be able to share the same `Container` declaration between both Swift versions of the library:
```

```swift

> #if compiler(>=5.5)

Reviewer 7: `compiler` is intended here and not `swift`?

Reviewer 2: So in the end we need `#if swift(>=5.5) && canImport(_Concurrency)`

public typealias MYPREFIX_Sendable = Swift.Sendable
#else 
public typealias MYPREFIX_Sendable = Any
#endif
```

> The Any alias is effectively a no-op when applied as generic constraint, and thus this way it is possible to keep the same Container<Value> declaration working across Swift versions.

Reviewer: ```suggestion
The `Any` alias is effectively a no-op when applied as generic constraint, and thus this way it is possible to keep the same `Container<Value>` declaration working across Swift versions.
```

### Task Local Values and Logging

> The newly introduced TaskLocal Values API ([SE-0311][SE-0311]) allows for implicit carrying of metadata along with Task execution. It is a natural fit for for tracing and carrying metadata around with task execution, and e.g. including it in log messages. 

Reviewer: ```suggestion
The newly introduced Task Local Values API ([SE-0311][SE-0311]) allows for implicit carrying of metadata along with `Task` execution. It is a natural fit for for tracing and carrying metadata around with task execution, and e.g. including it in log messages. 
```

> The newly introduced Task Local Values API ([SE-0311][SE-0311]) allows for implicit carrying of metadata along with `Task` execution. It is a natural fit for for tracing and carrying metadata around with task execution, and e.g. including it in log messages. 

Reviewer 7: ```suggestion
The newly introduced Task Local Values API ([SE-0311][SE-0311]) allows for implicit carrying of metadata along with `Task` execution. It is a natural fit for tracing and carrying metadata around with task execution, and e.g. including it in log messages. 
```

> We are working on adjusting Swift Log to become powerful enough to automatically pick up and log specific task local values. This change will be introduced in a source compatible way. 

Reviewer: ```suggestion
We are working on adjusting [SwiftLog](https://github.com/apple/swift-log) to become powerful enough to automatically pick up and log specific task local values. This change will be introduced in a source compatible way. 
```

> For now libraries should continue using logger metadata, but we expect a log of the manually passing metadata to each log statement can be replaced in the future with setting task local values. 

Reviewer 5: "a log of the manually passing metadata to each log statement" needs rephrasing

Reviewer 2: ah thanks, that's meant to be "a lot" :)

> For now libraries should continue using logger metadata, but we expect that a lot of the manually passing metadata to each log statement can be replaced in the future with setting task local values. 

Reviewer: ```suggestion
For now libraries should continue using logger metadata, but we expect that in the future a lot of the cases where metadata is manually passed to each log statement can be replaced with setting task local values. 
```

### Preparing for the concept of Deadlines

> Deadlines are another feature that closely relate to Swift Concurrency, and were originally pitched during the early versions of the Structured Concurrency proposal and later on moved out of it. The Swift team remains interested in introducing deadline concepts to the language and some preparation for it already has been performed inside the concurrency runtime. Right now however, there is no support for deadlines in Swift Concurrency and it is fine to continue using mechanisms like NIODeadline or similar mechanisms to cancel tasks after some period of time has passed. 

Reviewer: ```suggestion
Deadlines are another feature that closely relate to Swift Concurrency, and were originally pitched during the early versions of the Structured Concurrency proposal and later on moved out of it. The Swift team remains interested in introducing deadline concepts to the language and some preparation for it already has been performed inside the concurrency runtime. Right now however, there is no support for deadlines in Swift Concurrency and it is fine to continue using mechanisms like `NIODeadline` or similar mechanisms to cancel tasks after some period of time has passed. 
```

> Once Swift Concurrency gains deadline support, they will manifest in being able to cancel a task (and its child tasks) once such deadline (point in time) has been exceeded. For APIs to be “ready for deadlines” they don’t have to do anything special other than preparing to be able to deal with Tasks and their cancellation.

Reviewer: ```suggestion
Once Swift Concurrency gains deadline support, they will manifest in being able to cancel a task (and its child tasks) once such deadline (point in time) has been exceeded. For APIs to be “ready for deadlines” they don’t have to do anything special other than preparing to be able to deal with `Task`s and their cancellation.
```

### Cooperatively handling Task cancellation

> Task cancellation exists today in Swift Concurrency and is something that libraries may already handle. In practice it means that any asynchronous function (or function which is expected to be called from within Tasks), may use the Task.isCancelled or try Task.checkCancellation() APIs to check if the task it is executing in was cancelled, and if so, it may cooperatively abort any operation it was currently performing.

Reviewer: ```suggestion
`Task` cancellation exists today in Swift Concurrency and is something that libraries may already handle. In practice it means that any asynchronous function (or function which is expected to be called from within `Task`s), may use the [`Task.isCancelled`](https://developer.apple.com/documentation/swift/task/3814832-iscancelled) or [`try Task.checkCancellation()`](https://developer.apple.com/documentation/swift/task/3814826-checkcancellation) APIs to check if the task it is executing in was cancelled, and if so, it may cooperatively abort any operation it was currently performing.
```

> Cancellation can be useful in long running operations, or before kicking off some expensive operation. For example, an http client MAY check for cancellation before it sends an http request — it perhaps does not make sense to send a request if it is known the task awaiting on it does not care for the result anymore after all!

Reviewer 5: - "http client" -> "HTTP client" 
- "http request" -> "request" (enough, HTTP can be inferred)

Cancellation in general can be understood as “the one waiting for the result of this task is not interested in it anymore”, and it usually is best to throw a “cancelled” error when the cancellation is encountered. However, in some situations returning a “partial” result may also be appropriate (e.g. if a task is collecting many results, it may return those it managed to collect until now, rather than returning none or ignoring the cancellation and collecting all remaining results).

## What to expect with Swift 6

### Sendable: Global variables & imported code

Today, Swift 5.5 does not yet handle global variables at all within its concurrency checking model. This will soon change but the exact semantics are not set in stone yet. In general, avoid using global properties and variables wherever possible to avoid running into issues in the future. Consider deprecating global variables if able to.

> Some global variables have special properties, such as errno which contains the error code of system calls. It is a thread local variable and therefore safe to read from any thread/Task. We expect to improve the importer to annotate such globals with some kind of “known to be safe” annotation, such that the Swift code using it, even in fully checked concurrency mode won’t complain about it. Having that said, using errno and other “thread local” APIs is very error prone in Swift Concurrency because thread-hops may occur at any suspension point, so the following snippet is very likely incorrect:

Reviewer: ```suggestion
Some global variables have special properties, such as `errno` which contains the error code of system calls. It is a thread local variable and therefore safe to read from any thread/`Task`. We expect to improve the importer to annotate such globals with some kind of “known to be safe” annotation, such that the Swift code using it, even in fully checked concurrency mode won’t complain about it. Having that said, using `errno` and other “thread local” APIs is very error prone in Swift Concurrency because thread-hops may occur at any suspension point, so the following snippet is very likely incorrect:
```

```swift
sys_call(...)
await ...
let err = errno // BAD, we are most likely on a different thread here (!)
```

> Please take care when interacting with any thread-local API from Swift Concurrency. If your library had used thread local storage before, you will want to move them to use task-local values (https://github.com/apple/swift-evolution/blob/main/proposals/0311-task-locals.md) instead as they work correctly with Swift’s structured concurrency tasks.

Reviewer: ```suggestion
Please take care when interacting with any thread-local API from Swift Concurrency. If your library had used thread local storage before, you will want to move them to use [task-local values](https://github.com/apple/swift-evolution/blob/main/proposals/0311-task-locals.md) instead as they work correctly with Swift’s structured concurrency tasks.
```

Reviewer: not sure if it would make sense to include this with the other info on task local values above?

> Another tricky situation is with imported C code. There may be no good way to annotate the imported types as Sendable (or it would be too troublesome to do so by hand). Swift is likely to gain improved support for imported code and potentially allow ignoring some of the concurrency safety checks on imported code. 

Reviewer 5: Does the same apply to Objective-C interop? Worth mentioning?

Reviewer 2: The guide is for server ecosystem so there's no obj-c here :)

Reviewer: ```suggestion
Another tricky situation is with imported C code. There may be no good way to annotate the imported types as `Sendable` (or it would be too troublesome to do so by hand). Swift is likely to gain improved support for imported code and potentially allow ignoring some of the concurrency safety checks on imported code. 
```

> These relaxed semantics for imported code are not implemented yet, but keep it in mind when working with C APIs from Swift and trying to adopt the -warn-concurrency mode today. Please file any issues you hit so we can inform the development of these checking heuristics based on real issues you hit.

Reviewer: ```suggestion
These relaxed semantics for imported code are not implemented yet, but keep it in mind when working with C APIs from Swift and trying to adopt the `-warn-concurrency` mode today. Please file any issues you hit on [bugs.swift.org](https://bugs.swift.org/secure/Dashboard.jspa) so we can inform the development of these checking heuristics based on real issues you hit.
```

Reviewer 2: nice thanks

### Custom Executors

> We expect the ability that Swift Concurrency would allow custom executors. A custom executor would allow the ability to run actors / tasks “on” such executor. It is possible that EventLoops could become such executors, however the custom executor proposal has not been pitched yet.

Reviewer: ```suggestion
We expect that Swift Concurrency will allow custom executors in the future. A custom executor would allow the ability to run actors / tasks “on” such executor. It is possible that `EventLoop`s could become such executors, however the custom executor proposal has not been pitched yet.
```

While we expect potential performance gains from using custom executors “on the same event loop” by avoiding asynchronous hops between calls to different actors, their introduction will not fundamentally change how NIO libraries are structured.

The guidance here will evolve as Swift Evolution proposals for Custom Executors are proposed, but don’t hold off adopting Swift Concurrency until custom executors “land” - it is important to start adoption early. For most code we believe that the gains from adopting Swift Concurrency vastly outweigh the slight performance cost actor-hops might induce.

> ### Reduce use of Swift NIO Futures as “Concurrency Library“

Reviewer 7: ```suggestion
### Reduce use of SwiftNIO Futures as “Concurrency Library“
```

There are other occurrences of "Swift NIO" in this doc that should be SwiftNIO instead.

> Swift NIO currently provides a number of currency types for the Swift on Server ecosystem. Most notably `EventLoopFuture`s and `EventLoopPromise`s, that are used widely for asynchronous results. While the SSWG recommended using those in the past for easier interplay of server libraries, we advice to deprecate or remove them once Swift 6 lands. The swift-server ecosystem should go all in on the structured concurrency features the languages provides. For this reason, it is crucial to provide async/await APIs today, to give your library users time to adopt the new APIs.

Reviewer: ```suggestion
Swift NIO currently provides a number of currency types for the Swift on Server ecosystem. Most notably `EventLoopFuture`s and `EventLoopPromise`s, that are used widely for asynchronous results. While the SSWG recommended using those at the API level in the past for easier interplay of server libraries, we advise to deprecate or remove such APIs once Swift 6 lands. The swift-server ecosystem should go all in on the structured concurrency features the languages provides. For this reason, it is crucial to provide async/await APIs today, to give your library users time to adopt the new APIs.
```

> Swift NIO currently provides a number of currency types for the Swift on Server ecosystem. Most notably EventLoopFutures and EventLoopPromises, that are used widely for asynchronous results. While the sswg recommended using those in the past for easier interplay of server libraries, we advice to deprecate or remove them once Swift 6 lands. The swift-server ecosystem should go all in on the structured concurrency features the languages provides. For this reason, it is crucial to provide async/await APIs today, to give your library users time to adopt the new APIs.

Reviewer 5: "sswg" -> "SSWG"

> Some NIO types will remain however in the public interfaces of Swift on server libraries. We expect that networking clients and servers continue to be initialized with EventLoopGroups. The underlying transport mechanism (NIOPosix and NIOTransportServices) should become implementation details however and should not be exposed to library adopters.

Reviewer: ```suggestion
Some NIO types will remain however in the public interfaces of Swift on server libraries. We expect that networking clients and servers continue to be initialized with `EventLoopGroup`s. The underlying transport mechanism (`NIOPosix` and `NIOTransportServices`) should become implementation details however and should not be exposed to library adopters.
```

### NIO 3

> While subject to change, it is likely that Swift NIO will cut a 3.0 release in the months after Swift 6.0, at which point in time Swift will have enabled “full” Sendable checking.

Reviewer: ```suggestion
While subject to change, it is likely that Swift NIO will cut a 3.0 release in the months after Swift 6.0, at which point in time Swift will have enabled “full” `Sendable` checking.
```

> ### NIO 3

Reviewer 7: ```suggestion
### SwiftNIO 3
```

I suggest we use SwiftNIO throughout the doc.

Reviewer 2: Right, thanks Yim!

Do not expect NIO to suddenly become “more async”, NIO’s inherent design principles are about performing small tasks on the event loop and using Futures for any async operations. The design of NIO is not expected to change. It is crucial to its high performance networking design. Channel pipelines are not expected to become “async”.

> The NIO team will however use the chance to remove deprecated APIs and improve some APIs. The scope of changes should be comparable to the NIO1 → NIO2 version bump. If your SwiftNIO code compiles today with out warnings, chances are high that it will continue to work without modifications in NIO3.

Reviewer 5: "with out" -> "without"

After the release of NIO3, NIO2 will see bug fixes only.

### End-user code breakage

> It is expected that Swift 6 will break some code. As mentioned Swift NIO 3 is also going to be released sometime around Swift 6 dropping. Keeping this in mind, it might be a good idea to align major version releases around the same time, along with updating version requirements to Swift 6 and NIO 3 in your libraries.

Reviewer 6: I wonder if "swift 6 dropping" is a term that everyone will recognise.  eg, wondering what that would translate to if I ran this through google translate to a non-english language.

Reviewer 2: That's a good point, a bit too "slang-y" I guess, thanks! I'll reword

# Swift Concurrency adoption guidelines for Swift Server Libraries

This writeup attempts to provide a set of guidelines to follow by authors of server-side Swift libraries. Specifically a lot of the discussion here revolves around what to do about existing APIs and libraries making extensive use of Swift NIO’s `EventLoopFuture` and related types.

Swift Concurrency is a multi-year effort. It is very valuable for the server community to participate in this multi-year adoption of the concurrency features, one by one, and provide feedback while doing so. As such, we should not hold off adopting concurrency features until Swift 6 as we may miss valuable opportunity to improve the concurrency model.

In 2021 we saw structured concurrency and actors arrive with Swift 5.5. Now is a great time to provide APIs using those primitives. In the future we will see fully checked Swift concurrency. This will come with breaking changes. For this reason adopting the new concurrency features can be split into two phases.

## What you can do right now

> ### `#if` guarding code using Concurrency

Reviewer 3: The subsections are roughly categorised as API, implementation and preparing for the future, so I think the "API design" and "SwiftNIO Helper Functions" sections should come before this one.

Reviewer 2: done thanks

In order to have code using concurrency along with code not using concurrency, you may have to `#if` guard certain pieces of code. The correct way to do so is the following:

```swift

> #if compiler(>=5.5) && canImport(_Concurrency)

Reviewer 3: There are bunch of places in this guide which are missing `&& canImport(_Concurrency)`. There are also places where `swift(>=5.5)` instead of `compiler(>=5.5)`.

Reviewer 3: I think it's worth adding a note about `@available` requirements too.

Reviewer 2: Thanks done

...
#endif
```

Please note that you do _not_ need to _import_ the `_Concurrency` at all, if it is present it is imported automatically.

```swift

> // DO NOT DO THIS.
> // Instead don't do any import and it'll import automatically when possible.
> #if compiler(>=5.5) && canImport(_Concurrency)
> import _Concurrency
> #endif

Reviewer 3: nit: make it even-more-obvious:

```suggestion
#if compiler(>=5.5) && canImport(_Concurrency)
// DO NOT DO THIS.
// Instead don't do any import and it'll import automatically when possible.
import _Concurrency
#endif
```

```

### API Design

Firstly, existing libraries should strive to add `async` functions where possible to their user-facing “surface” APIs in addition to existing `*Future` based APIs wherever possible. These additive APIs can be gated on the Swift version and can be added without breaking existing users' code, for example like this:

```swift
extension Worker {
  func work() -> EventLoopFuture<Value> { ... }
  
  #if swift(>=5.5)
  func work() async throws -> Value { ... }
  #endif
}
```

If a function cannot fail but was using futures before, it should not include the `throws` keyword in its new incarnation. 

> Such adoption can begin immediately, and should not cause any issues to existing users of existing libraries. 

Reviewer 3: I think there are broader changes that libraries can make as well, e.g. `AsyncSequence` based APIs rather than delegate based APIs, for example.

It may also be worth calling out that given we're on the road to Swift 6, libraries and the ecosystem as a whole are going to have to go through major version bumps so adding async-await APIs now can help lessen the pain of going through a major version change later because new APIs can be adopted incrementally.

### SwiftNIO helper functions

To allow an easy transition to async code, SwiftNIO offers a number of helper methods on `EventLoopFuture` and `-Promise`. Those live in the `_NIOConcurrency` module and will move to `NIOCore` once Swift concurrency is released.

On every `EventLoopFuture` you can call `.get()` to transition the future into an `await`-able invocation. If you want to translate async/await calls to an `EventLoopFuture` we recommend the following pattern:

```swift 
#if swift(>=5.5)
func yourAsyncFunctionConvertedToAFuture(on eventLoop: EventLoop) 
    -> EventLoopFuture<Result> {
    let promise = context.eventLoop.makePromise(of: Out.self)
    promise.completeWithTask {

>         try await yourMethod(yourInputs)

Reviewer 8: Question about this; from what I can tell, this won't run `yourMethod` on an `EventLoop`; (assuming because I tried an example and `MultiThreadedEventLoopGroup.currentEventLoop` is nil).

Is there a way to have async tasks run on an event loop?

I ask since I thought that ideally during a request's lifecycle it won't be hopping off the EventLoop it's received on, even if running async tasks like a database lookup.

I might be confused on what's going on here too 🙂

Reviewer 2: > Is there a way to have async tasks run on an event loop?

Not today; that's what upcoming work on custom executors in Swift Concurrency would allow.

Reviewer 8: The specific use case is that when running database queries, I've got a dict of connection pools, one for each `EventLoop`.

When a query is made, a pool is found using the current event loop's objectidentifier. But it seems like with async functions, they aren't running on an event loop and so if they need to run a database query it's not possible to look up a pool based on that.

Reviewer 2: that's correct

### Custom Executors

We expect that Swift Concurrency will allow custom executors in the future. A custom executor would allow the ability to run actors / tasks “on” such executor. It is possible that `EventLoop`s could become such executors, however the custom executor proposal has not been pitched yet.

While we expect potential performance gains from using custom executors “on the same event loop” by avoiding asynchronous hops between calls to different actors, their introduction will not fundamentally change how NIO libraries are structured.

The guidance here will evolve as Swift Evolution proposals for Custom Executors are proposed, but don’t hold off adopting Swift Concurrency until custom executors “land” - it is important to start adoption early. For most code we believe that the gains from adopting Swift Concurrency vastly outweigh the slight performance cost actor-hops might induce.

### Reduce use of Swift NIO Futures as “Concurrency Library“

> Swift NIO currently provides a number of currency types for the Swift on Server ecosystem. Most notably `EventLoopFuture`s and `EventLoopPromise`s, that are used widely for asynchronous results. While the SSWG recommended using those at the API level in the past for easier interplay of server libraries, we advise to deprecate or remove such APIs once Swift 6 lands. The swift-server ecosystem should go all in on the structured concurrency features the languages provides. For this reason, it is crucial to provide async/await APIs today, to give your library users time to adopt the new APIs.

Reviewer 9: ```suggestion
Swift NIO currently provides a number of concurrency types for the Swift on Server ecosystem. Most notably `EventLoopFuture`s and `EventLoopPromise`s, that are used widely for asynchronous results. While the SSWG recommended using those at the API level in the past for easier interplay of server libraries, we advise to deprecate or remove such APIs once Swift 6 lands. The swift-server ecosystem should go all in on the structured concurrency features the languages provides. For this reason, it is crucial to provide async/await APIs today, to give your library users time to adopt the new APIs.
```

> Some NIO types will remain however in the public interfaces of Swift on server libraries. We expect that networking clients and servers continue to be initialized with `EventLoopGroup`s. The underlying transport mechanism (`NIOPosix` and `NIOTransportServices`) should become implementation details however and should not be exposed to library adopters.

Reviewer 9: Once custom executors are introduced, will passing in `EventLoopGroups` to libraries still be necessary? Or will an application author be able to set it once and tasks created by libraries will automatically be executed on it?

Reviewer 10: The answer depends very much on the specifics of the custom executor proposal.

### NIO 3

While subject to change, it is likely that Swift NIO will cut a 3.0 release in the months after Swift 6.0, at which point in time Swift will have enabled “full” `Sendable` checking.

> Do not expect NIO to suddenly become “more async”, NIO’s inherent design principles are about performing small tasks on the event loop and using Futures for any async operations. The design of NIO is not expected to change. It is crucial to its high performance networking design. Channel pipelines are not expected to become “async”.

Reviewer 9: >The design of NIO is not expected to change. It is crucial to its high performance networking design.

This line seems to imply that `async` cannot be used without significant overhead, is this the case? If so, should we include guidance in this document about when _not_ to completely adopt `async`, as NIO has opted not to? Or maybe a note saying NIO is a special case due to its low level nature, and that most/all other libraries should try to become more `async`?

The reason I mention this is that the majority of the document focuses on encouraging adoption of `async`, but at the very end here we have a note about how NIO will largely not be doing so, which may be a bit contradictory. A little bit further clarification could help with that though.

Reviewer 10: NIO is not doing so because of the concerns around the “bottom” of the `async` pyramid. `async`/`await` is built on top of a co-operative thread pool within the Swift runtime. This thread pool must not be blocked by any operation, because doing so will starve the pool and prevent further progress of other async tasks.

This poses a problem for I/O. At some point, someone somewhere has to block a thread waiting for more I/O events, either in an I/O syscall or in something like `epoll_wait`. This is how NIO works: each of the event loop threads ultimately blocks on `epoll_wait`. We can’t do that inside the cooperative thread pool, as to do so would starve it for other async tasks, so we’d have to do so on a different thread.

This would mean we’d have a thread-hop between each I/O operation and dispatching it onto the async/await pool. This is not acceptable for high performance I/O: the context switch for each I/O op is too expensive. As a result, until we can safely block a thread for ourselves (with concurrent executors) we simply cannot use the async/await pattern in NIO code.

The rule of thumb is: if you need the absolute fastest computation on data from the network, you write `ChannelHandler`s and do your I/O on event loops. If you can tolerate slightly less performance there, likely because your business logic is more substantial, then async/await will repay you in ease of use much more than the (small) amount it’ll cost you in perf.

Reviewer 9: That all makes sense, thanks for the detailed answer! I think it may be beneficial then to reword the existing paragraph to include some of this context and to be bit more explanatory rather than declarative (i.e. "NIO's API will largely remain the same because..." rather than "don't expect NIO's API to change..."). This will also help library authors decide when they might need to use NIO's facilities more directly rather than higher level `async`/`await` ones.

Reviewer 2: Yeah, happy to take some of the wording into the doc 👍

The NIO team will however use the chance to remove deprecated APIs and improve some APIs. The scope of changes should be comparable to the NIO1 → NIO2 version bump. If your SwiftNIO code compiles today without warnings, chances are high that it will continue to work without modifications in NIO3.

> After the release of NIO3, NIO2 will see bug fixes only.

Reviewer 3: We'll fix security vulnerabilities too!

Reviewer 2: thanks will add

### End-user code breakage

It is expected that Swift 6 will break some code. As mentioned Swift NIO 3 is also going to be released sometime around Swift 6 dropping. Keeping this in mind, it might be a good idea to align major version releases around the same time, along with updating version requirements to Swift 6 and NIO 3 in your libraries.

Both Swift and Swift NIO are not planning to do “vast amounts of change”, so adoption should be possible without major pains.

### Guidance for library users

> As soon as Swift 6 comes out, we recommend using the latest Swift 6 toolchains, even if using the Swift 5.5.n language mode (which may yield only warnings rather than hard failures on failed Sendability checks).

Reviewer 3: Worth saying why we make this recommendation

Reviewer 2: thanks

> As soon as Swift 6 comes out, we recommend using the latest Swift 6 toolchains, even if using the Swift 5.5.n language mode (which may yield only warnings rather than hard failures on failed Sendability checks). This will result in better warnings and compiler hints, than just using a 5.5 toolchain.
